From 1d2b51b0bd8c587af933a9049a59ffbe7145e214 Mon Sep 17 00:00:00 2001
From: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
Date: Thu, 7 May 2026 12:44:40 +0530
Subject: [PATCH] drivers: bus: mhi: host: Add Debug Reg dump in RDDM Panic
 flow

Add debug register dump in RDDM flow invoked as part of kernel panic
handler. This would provide breadcrumbs to debug RDDM failure in kernel
panic flow.

Change-Id: Ied92d9366626f1f61d15764a8bc60ebd5b2d71c5
Signed-off-by: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
---
 drivers/bus/mhi/host/boot.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/bus/mhi/host/boot.c b/drivers/bus/mhi/host/boot.c
index 4a37e3de96a2..b5305d388f4d 100644
--- a/drivers/bus/mhi/host/boot.c
+++ b/drivers/bus/mhi/host/boot.c
@@ -101,6 +101,7 @@ static int __mhi_download_rddm_in_panic(struct mhi_controller *mhi_cntrl)
 	u32 rx_status;
 	enum mhi_ee_type ee;
 	const u32 delayus = 2000;
+	const u32 soc_reset_delay_ms = 200;
 	u32 retry = (mhi_cntrl->timeout_ms * 1000) / delayus;
 	const u32 rddm_timeout_us = 400000;
 	int rddm_retry = rddm_timeout_us / delayus;
@@ -153,11 +154,18 @@ static int __mhi_download_rddm_in_panic(struct mhi_controller *mhi_cntrl)
 			/* Hardware reset so force device to enter RDDM */
 			dev_dbg(dev,
 				"Did not enter RDDM, do a host req reset\n");
+			mhi_debug_reg_dump(mhi_cntrl);
 			mhi_soc_reset(mhi_cntrl);
-			udelay(delayus);
+			mdelay(soc_reset_delay_ms);
 		}
 
 		ee = mhi_get_exec_env(mhi_cntrl);
+		/* If Target still did not swich to RDDM, dump debug registers and bail out */
+		if (ee != MHI_EE_RDDM) {
+			dev_err(dev, "Failed to switch to RDDM\n");
+			mhi_debug_reg_dump(mhi_cntrl);
+			return -EIO;
+		}
 	}
 
 	dev_dbg(dev,
-- 
2.34.1

